### PR TITLE
Add Jack compiler

### LINKED ISSUE #23: Jack Language
Implement parser for the Jack language

### PR DESCRIPTION
**Pull request by netalondon**
Closes #23

### CODE REVIEW

File: components/src/stores/compiler.store.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    setSelected(state: CompilerPageState, selected: string) {
+      state.selected = selected;
+    },
+  };
+
+  const actions = {
+    async editFile(name: string, content: string) {
+      dispatch.current({ action: "setFile", payload: { name, content } });
+      await fs.writeFile(`/jack/${name}.jack`, content);
```
Comment (DavidSouther) [inline components/src/stores/compiler.store.ts:78]: This gets repeated a lot, maybe a `makeJackFilename` helper?

File: simulator/src/jack/compiler.ts [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+  varData(name: string): VariableData | undefined {
+    return this.localSymbolTable[name] || this.globalSymbolTable[name];
+  }
+
+  var(name: string) {
+    const data = this.varData(name);
+    return `${data?.segment} ${data?.index}`;
+  }
+
+  write(line: string) {
```
Comment (DavidSouther) [inline simulator/src/jack/compiler.ts:88]: I'd just make this 

```
write(...lines: string[]) {
  this.instructions.push(...lines);
}

this.write("add");
this.write(...expression);
```

File: simulator/src/jack/compiler.ts [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+  }
+
+  compileSubroutineCall(call: SubroutineCall) {
+    let object = "";
+    let className = "";
+    let subroutineName = "";
+    let isMethod = true;
+
+    if (call.name.includes(".")) {
+      const [prefix, suffix] = call.name.split(".");
```
Comment (DavidSouther) [inline simulator/src/jack/compiler.ts:260]: `.split(".", 2)` might be a good safety against an "accidental" `foo.bar.baz` name.

File: web/src/pages/compiler.tsx [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        <div
+          role="tablist"
+          style={
+            {
+              "--tab-count": `${Object.keys(state.files).length}`,
+            } as CSSProperties
+          }
+        >
+          {Object.keys(state.files).map((file) => (
+            <>
```
Comment (DavidSouther) [inline web/src/pages/compiler.tsx:190]: I just made a <TabList> and <Tab> helper - https://github.com/DavidSouther/ailly/blob/main/web/src/components/Tabs.tsx

We could copy that over to here and make this much simpler.

File: web/src/urls.tsx
Comment (DavidSouther) [inline web/src/urls.tsx:1]: Reminder to pull this into a separate PR when it's time to release.

**Review by DavidSouther — COMMENTED** [review]
Looks like a really great start!

Some initial thoughts, looking forward to seeing it entirely in action!

File: web/src/shell/tabs.tsx
Comment (DavidSouther) [inline web/src/shell/tabs.tsx:1]: Could we pull this into its own PR and use it for the test panel first? This this PR only needs to use it

File: simulator/src/languages/jack.ts
Comment (DavidSouther) [inline simulator/src/languages/jack.ts:1]: Some tests for parsing at least the example Jack files

File: simulator/src/jack/compiler.ts
Comment (DavidSouther) [inline simulator/src/jack/compiler.ts:1]: Some compiler tests, please

**Review by DavidSouther — CHANGES_REQUESTED** [review]

File: simulator/src/languages/jack.ts
Comment (netalondon) [reply simulator/src/languages/jack.ts:1]: Is there a better approach here than to generate the expected cases for each test programmatically?

File: simulator/src/languages/jackSamples.ts
Comment (DavidSouther) [inline simulator/src/languages/jackSamples.ts:1]: We've usually been putting these in https://github.com/nand2tetris/web-ide/tree/main/projects/src/samples

**Review by DavidSouther — CHANGES_REQUESTED** [review]

File: simulator/src/languages/jack.ts
Comment (DavidSouther) [reply simulator/src/languages/jack.ts:1]: No, this is good for now. As we run into bugs or errors, focused tests can be very helpful to troubleshoot and isolate, but for now "Parse the sample files" is a good baseline.

**Review by DavidSouther — APPROVED** [review]

**Review by DavidSouther — APPROVED** [review]

**Pull request merged**